File: restore/server/src/main/java/bgu/spl/net/impl/Frames/ConnectFrame.java
package bgu.spl.net.impl.Frames;

import bgu.spl.net.impl.User.User;
import bgu.spl.net.impl.gamesUpdate.GamesFeed;
import bgu.spl.net.impl.stomp.StompCommand;
import bgu.spl.net.impl.stomp.StompFrame;

import java.util.HashMap;
import java.util.Map;

public class ConnectFrame extends StompFrame{
    String acceptVersion;
    String host;
    String login;
    String passcode;

    public ConnectFrame(Map<String,String> headers) {
        super(StompCommand.CONNECT, headers);
    }

    @Override 
    public StompFrame create(String[] body){
         if(!stompsHeaders.containsKey("accept-version")){
            missingHeaders.add("accept-version");
            errorAccourd = true;
        }
        if(!stompsHeaders.containsKey("host")){
            missingHeaders.add("host");
            errorAccourd = true;
            
        }
        if(!stompsHeaders.containsKey("login")){
           missingHeaders.add("login");
           errorAccourd = true;
        }
        if(!stompsHeaders.containsKey("passcode")){
           missingHeaders.add("passcode");
           errorAccourd = true;
        }
        if(stompsHeaders.containsKey("receipt-id")){
            receiptId = stompsHeaders.get("receipt-id");
            errorHeaders.put("receipt-id", receiptId);
       }

       if(errorAccourd){
            return frameUtil.buildHeaderMissingErrorFrame(errorHeaders,missingHeaders);
       }

        login = stompsHeaders.get("login");
        passcode = stompsHeaders.get("passcode");
        acceptVersion = stompsHeaders.get("accept-version");
        host = stompsHeaders.get("host");
        return this;
    }


    @Override
    public void excute(GamesFeed gamesFeed, User user){
        Map<String,User> users = gamesFeed.allUsers;
        String errorStr="";
        errorAccourd = false;
        
        // new connection
        if(user.getUsername()==null){
            user.setDetails(login, passcode);
        }

        //checks if the password is worng or if this user is already logged in
        // and sends an Error frame 
        if(users.keySet().contains(login)){
            User currUser = users.get(login);
            if(currUser.getIsConnected()){
                errorAccourd = true;
                errorStr = "User already logged in";
            }
            else if(user.getPassword() != currUser.getPassword()){
                errorStr = "Wrong password";
                errorAccourd = true;
            }
        }

        receiptId = null;
        if(stompsHeaders.containsKey("receipt")){
            receiptId = stompsHeaders.get("receipt");
        }

        if(errorAccourd){
            StompFrame error = frameUtil.buildErrorFrameForCases(receiptId, errorStr);
            error.excute(gamesFeed, user);
            return;
        }

        user.connect();
        gamesFeed.allUsers.put(login,user); // move to connectFrame?
        StompFrame msg = buildConnectedFrame();
        // excute the connected frame that will be send to the connected user
        msg.excute(gamesFeed, user);
    }

    private StompFrame buildConnectedFrame(){
        Map<String,String> headers = new HashMap<String,String>();
        headers.put("version", "1.2");

        return frameUtil.createFrame(StompCommand.CONNECTED,headers,null);
    }
}

File: restore/client/src/Frames/connectFrame.cpp
#include "../include/StompFrame.h"
#include <iostream>
using std::cout;

//we send to the server
ConnectFrame::ConnectFrame(): mVersion("1.2"), mHost(""), mLogin(""), mPasscode(""){
    command = CONNECT;
}
ConnectFrame::ConnectFrame(string host, int port, string login, string passcode): mVersion("1.2"), mHost(host), mLogin(login),mPasscode(passcode){
    command = CONNECT;
    host_ = host;
    port_ = port;
    headers.emplace("login", login);
    headers.emplace("passcode", passcode);
    headers.emplace("host", host);
    headers.emplace("accept-version", "1.2");
}

ConnectFrame::~ConnectFrame(){}


void ConnectFrame::excute(user* curUser){
    // convert the frame to string  
    string sendFrame = toString();
    //sends the string to the server 
    curUser->send(sendFrame, '\0');
}

void ConnectFrame::create(map<string,string> headers, string body ){
    if (headers.find("host") == headers.end()) {
        cout << "error - host not found"; // todo: remove
    } else if(headers.find("login") == headers.end()){
        cout << "error - login not found"; // todo: remove
    } else if(headers.find("passcode") == headers.end()){
        cout << "error - passcode not found"; // todo: remove)
    } else {
        mHost = headers["host"];
        mLogin = headers["login"];
        mPasscode = headers["passcode"];
    }
}